### PR TITLE
[IMP] account: easier payment registration for bills

### PR DESCRIPTION
**Pull request by Megaaaaaa**
Description of the issue/feature this commit addresses:

Currently, the payment registration for vendor bills is kinda hard to figure. Noticing which bills are overdue, which require a payment is not made so easy by the interface and could be improved. Also there is no way to register a payment for multiple lines at a time which could improve the UX.

---

Desired behavior after the commit is merged :

On the dashboard, all the buttons in the "Customer Invoices" and "Vendor Bills" have been renamed to remove the "Invoices" and the "Bills" inside them.

The buttons "Bills to Pay" and "Late Bills" which now are "To Pay" and "Late" (see above) redirect to the Journal Items with specific filters to show only the necessary lines.

The order of the columns in the Journal Items has been changed and their visibility has been changed too.

The default filters of the Journal Items when coming from the dashboard buttons has been changed. Payable has been replaced by Purchases.

Filters, Group by and Search by options have been added for the Due Date and the Discount Date fields in the Journal Items.

---

task-3516496

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_journal_dashboard.py [hunk 1]
```diff
@@ -663,6 +713,29 @@ def action_create_vendor_bill(self):
...
+        return {
+            'name': _('Payment Items'),
+            'type': 'ir.actions.act_window',
+            'res_model': 'account.move.line',
+            'views': [[view_id, 'tree']],
+            'view_mode': 'tree',
+            'search_view_id': [search_view_id, 'search'],
+            'target': 'current',
+            'context': {'search_default_group_by_partner':1, 'expand': 1},
+            'domain': [('parent_state', '=', 'posted'), ('amount_residual', '!=', 0), ('account_id.reconcile', '=', True)],
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:728]: The point of putting those in the domain instead of using the already existing filters was for the search bar not to be filled with filters the user will never want to remove and therefore don't need to be displayed.
I couldn't figure a way to hide filters so I put those in the domain.

File: addons/account/views/account_journal_dashboard_view.xml [hunk 2]
```diff
@@ -330,14 +330,13 @@
...
                                     context="{'search_default_open':1, 'search_default_posted':1, 'search_default_partial': 1}" id="account_dashboard_sale_pay_link">
-                                        <t t-out="dashboard.number_waiting"/> Unpaid Invoices
+                                        <span t-if="journal_type == 'sale'" title="Unpaid Invoices"><t t-out="dashboard.number_waiting"/> Unpaid Invoices</span>
                                     </a>
-
-                                    <a type="object" t-if="journal_type == 'purchase'" name="open_action"
-                                    context="{'search_default_open':1, 'search_default_posted':1, 'search_default_partial': 1}" id="account_dashboard_purchase_pay_link">
-                                        <t t-out="dashboard.number_waiting"/> Bills to Pay
+                                    <a type="object" name="action_open_payment_items"
+                                    context="{'search_default_purchases': '1'}">
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/views/account_journal_dashboard_view.xml:338]: The "Invoices to Validate", "Bills to Validate", "Unpaid Invoice" and "Late Invoice" texts are clickable and trigger the open_action method. That method assigns the right filters and domain according to the text that was clicked. However, all those redirect to an account.move view while "Bills to Pay" and "Late Bills" redirect to an account.move.line view. Using the same method would probably involve a lot of changes while using a different method is way clearer. This new method is action_open_payment_items and return an easy action which needs a simple domain change depending on whether you click on one or the other.

In short : Sales is used when clicking on an invoice related text but it is done inside the action instead of passed via the context.

File: addons/web/static/src/search/search_bar/search_bar.js
Comment (william-andre) [inline addons/web/static/src/search/search_bar/search_bar.js:1]: @odoo/rd-framework-js Are you ok with this?

The purpose is to be able to have a quick search that searches before a certain date by default (see `_search_payment_date`)

File: addons/account/views/account_move_views.xml [hunk 3]
```diff
@@ -1582,6 +1669,23 @@
             </field>
         </record>
 
+        <record id="action_open_payment_items" model="ir.actions.act_window">
```
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/views/account_move_views.xml:1668]: You resolved but it is still here?

File: addons/account/models/account_journal_dashboard.py [hunk 4]
```diff
@@ -664,6 +714,56 @@ def action_create_vendor_bill(self):
...
+            elif model_info[0] == 'account.move.line':
+                action.get('domain').append(('move_name', '=', self.env[model_info[0]].search([('id', '=', model_info[1])]).move_name))
+
+            if options.get('period'):
+                action.get('domain').extend(
+                    self.build_domain_from_period(options.get('period'))
+                )
+        return action
+
+    def build_domain_from_period(self, period):
```
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:754]: Same, should be in `account_reports`

File: addons/account/views/account_journal_dashboard_view.xml [hunk 5]
```diff
@@ -320,8 +320,8 @@
                             <div class="row" t-if="dashboard.number_draft">
                                 <div class="col overflow-hidden text-start">
                                     <a type="object" name="open_action" context="{'search_default_draft': '1'}">
-                                        <span t-if="journal_type == 'sale'" title="Invoices to Validate"><t t-out="dashboard.number_draft"/> Invoices to Validate</span>
-                                        <span t-if="journal_type == 'purchase'" title="Bills to Validate" id="account_dashboard_purchase_draft"><t t-out="dashboard.number_draft"/> Bills to Validate</span>
+                                        <span t-if="journal_type == 'sale'"><t t-out="dashboard.number_draft"/> Invoices to Validate</span>
+                                        <span t-if="journal_type == 'purchase'" id="account_dashboard_purchase_draft"><t t-out="dashboard.number_draft"/> Bills to Validate</span>
```
Comment (william-andre) [inline addons/account/views/account_journal_dashboard_view.xml:324]: Is there a reason to remove the `title`?
I think it can be useful on smaller screens, for instance:
image

Could be even worse with some languages

File: addons/account/views/account_journal_dashboard_view.xml [hunk 6]
```diff
@@ -332,12 +332,11 @@
                                 <div class="col overflow-hidden text-start">
                                     <a type="object" t-if="journal_type == 'sale'" name="open_action"
                                     context="{'search_default_open':1, 'search_default_posted':1, 'search_default_partial': 1}" id="account_dashboard_sale_pay_link">
-                                        <t t-out="dashboard.number_waiting"/> Unpaid Invoices
+                                        <span><t t-out="dashboard.number_waiting"/> Unpaid Invoices</span>
```
Comment (william-andre) [inline addons/account/views/account_journal_dashboard_view.xml:335]: Any reason to add a `span`?

**Review by william-andre — COMMENTED** [review]
It's getting there! :smile:

File: addons/account/views/account_move_views.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/views/account_move_views.xml:1668]: I thougth your first comment was about the action_open_payment_item() method which was present in both the odoo/addons/account/models/account_journal_dashboard.py and enterprise/account_reports/models/account_aged_partner_balance.py and was indeed a duplicate. Wasn't it ?

This action_open_payment_item is the menu item in the drop-down when clicking on "Vendors" in the bar at the top of the UI. Am i not forced to create a specific action for such a button ?

File: addons/account/models/account_journal_dashboard.py [hunk 7]
```diff
@@ -664,6 +714,56 @@ def action_create_vendor_bill(self):
...
+                    """) % (
+                        _("Amounts to settle"),
+                        _("Cool, it looks like you don't have anything to pay either for vendor bills or for customer credit notes.")
+            ),
+        }
+        if options:
+            if options.get('context'):
+                action.get('context').update(options.get('context'))
+
+            model_info = self.env['account.report']._get_model_info_from_id(options['line']['id'])
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:742]: You want me to duplicate _get_model_info_from_id() in this model ?

File: addons/account/views/account_journal_dashboard_view.xml (hunk 5 above)
Comment (Megaaaaaa) [reply addons/account/views/account_journal_dashboard_view.xml:324]: I removed them after this this comment in your first review: https://github.com/odoo/odoo/pull/138188#discussion_r1360590483 . 

I'm putting them back waiting for your confirmation but I think it is indeed better to keep them.

Comment (william-andre) [reply addons/account/views/account_journal_dashboard_view.xml:324]: Yes, I probably got confused in the diff before, which was more complicated

File: addons/account/models/account_journal_dashboard.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/models/account_journal_dashboard.py:742]: No, I want you to extend this method in `account_reports` to use `_get_model_info_from_id` over there since it doesn't exist in `account`

File: addons/account/views/account_move_views.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/views/account_move_views.xml:1668]: My issue is that there is a lot of duplicated data

File: addons/account/models/account_journal_dashboard.py [hunk 8]
```diff
@@ -437,17 +439,45 @@ def _fill_sale_purchase_dashboard_data(self, dashboard_data):
...
-        query_results_drafts = group_by_journal(self.env.cr.dictfetchall())
+        query_results_bills_to_pay = group_by_journal(self.env.cr.dictfetchall())
+        if purchase_journals:
+            query, params = purchase_journals._get_open_payments_query().select(*payment_field_list)
+            self.env.cr.execute(query, params)
+            query_results_payments_to_pay = group_by_journal(self.env.cr.dictfetchall())
+        query_results_to_pay = {}
+        for journal in sale_purchase_journals:
+            if journal.type == "purchase":
+                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id] or query_results_bills_to_pay[journal.id]
```
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:463]: Well, it does. For some journals, it seems that the aml query doesn't get the right totals while the move one does. I probably don't know enough about specific use cases to understand why but a test has been written for such a case and without that backup, the sum_waiting and sum_late end up being wrong.

File: addons/account/views/account_move_views.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/views/account_move_views.xml:1668]: Could you tell me how to use an existing action as a menuitem then ?

[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/views/account_move_views.xml:1668]: That you can't do.
On the other hand, you can read the xml-defined action in python and reuse it like this
```python
action = self.env['ir.actions.actions']._for_xml_id('account.action_open_payment_items')
```

File: addons/web/static/src/search/search_bar/search_bar.js
Comment (sdegueldre) [reply addons/web/static/src/search/search_bar/search_bar.js:1]: @Polymorphe57

File: addons/web/static/src/search/search_bar/search_bar.js [hunk 9]
```diff
@@ -236,6 +231,21 @@ export class SearchBar extends Component {
         return items;
     }
 
+    getPreposition(searchItem) {
+        const [, fieldType] = this.getFieldInfos(searchItem);
+        return _t(["date", "datetime"].includes(fieldType) ? "at" : "for");
```
Comment (sdegueldre) [inline addons/web/static/src/search/search_bar/search_bar.js:236]: I know this line was just moved but if this works it works by accident, the terms cannot be extracted from this, they should be translated individually
```suggestion
        return ["date", "datetime"].includes(fieldType) ? _t("at") : _t("for");
```

File: addons/web/static/src/search/search_bar/search_bar.js [hunk 10]
```diff
@@ -236,6 +231,21 @@ export class SearchBar extends Component {
...
+    }
+
+    getFieldInfos(searchItem) {
+        const isFieldProperty = searchItem.type === "field_property";
+        const { type } = isFieldProperty
+            ? searchItem.propertyFieldDefinition
+            : this.fields[searchItem.fieldName];
+        const fieldType = type === "reference" ? "char" : type;
+
+        return [isFieldProperty, fieldType]
```
Comment (william-andre) [inline addons/web/static/src/search/search_bar/search_bar.js:246]: Isn't this cleaner? To potentially avoid playing with indexes
```suggestion
        return {isFieldProperty, fieldType}
```

File: addons/account/models/account_journal_dashboard.py [hunk 11]
```diff
@@ -437,17 +440,53 @@ def _fill_sale_purchase_dashboard_data(self, dashboard_data):
...
+            query, params = purchase_journals._get_open_bills_to_pay_query().select(*bills_field_list)
+            self.env.cr.execute(query, params)
+            query_results_bills_to_pay = group_by_journal(self.env.cr.dictfetchall())
+            query, params = purchase_journals._get_open_payments_query().select(*payment_field_list)
+            self.env.cr.execute(query, params)
+            query_results_payments_to_pay = group_by_journal(self.env.cr.dictfetchall())
+        else:
+            query, params = sale_journals._get_open_bills_to_pay_query().select(*bills_field_list)
+            self.env.cr.execute(query, params)
+            query_results_bills_to_pay = group_by_journal(self.env.cr.dictfetchall())
```
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:465]: So you don't compute anything for the sales journals if you have purchase journals?

File: addons/account/models/account_move_line.py [hunk 12]
```diff
@@ -1121,6 +1139,33 @@ def _compute_analytic_distribution(self):
...
                 line.analytic_distribution = distribution or line.analytic_distribution
 
+    @api.depends('discount_date', 'date_maturity')
+    def _compute_payment_date(self):
+        for line in self:
+            line.payment_date = line.discount_date if line.discount_date and date.today() <= line.discount_date else line.date_maturity
+
+    def _search_payment_date(self, operator, value):
+        assert operator == '='
+        return ['|', '&', ('discount_date', '>=', str(date.today())), ('discount_date', '<=', value), ('date_maturity', '<=', value)]
```
Comment (william-andre) [inline addons/account/models/account_move_line.py:1149]:
```suggestion
        return [
            '|', 
            '&', ('discount_date', '>=', str(date.today())), ('discount_date', '<=', value), 
            '&', ('discount_date', '<', str(date.today())), ('date_maturity', '<=', value),
        ]
```

We need to search on either one either the other field, so the clauses on today need to be opposite to each other.
Also need to check if it is working as expected when `discount_date` is not set?

Also, add a test case for this (create multiple lines with various dates), and do multiple searches with mocked today (use `freeze_time`)

File: addons/account/models/account_journal_dashboard.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (william-andre) [reply addons/account/models/account_journal_dashboard.py:468]: > it seems that the aml query doesn't get the right totals while the move one does

What if the query returns an amount that is not right but that is not zero then?

It's the concept of the `or` that makes me think it has to be wrong. It is right or it is not. Maybe you might need to sum both or something like that?

Anyways, needs a test.

File: addons/account/models/account_journal_dashboard.py [hunk 13]
```diff
@@ -664,6 +723,52 @@ def action_create_vendor_bill(self):
...
+                domain += options_domain
+            model_info = self._get_model_info_from_id(options['line']['id'])
+            if model_info[0] == 'res.partner':
+                domain.append(('partner_id', '=', model_info[1]))
+            elif model_info[0] == 'account.move.line':
+                domain.append(('move_name', '=', self.env[model_info[0]].search([('id', '=', model_info[1])]).move_name))
+
+            if options.get('period'):
+                domain.extend(
+                    self.env['account.report'].build_domain_from_period(options.get('period'))
```
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:741]: This is still not defined in community

Does any extension from here (the `options` management) actually need to be in community?
It all seems only related to `account_reports`, keep it in `account_reports`

File: addons/account/models/account_journal_dashboard.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Megaaaaaa) [reply addons/account/models/account_journal_dashboard.py:468]: If the query retrieved a value, it is right. If it didn't, the other query got it right. Depending on the journal, the aml query will retrieve something or nothing at all. The final result is not some pieces of information coming from one query and some from the other. It is either one or the other, there is no "merge" of data with that `or`. 
I could replace:
```py
for journal in sale_purchase_journals:
            if journal.type == "purchase":
                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id] or query_results_bills_to_pay[journal.id]
            else:
                query_results_to_pay[journal.id] = query_results_bills_to_pay[journal.id]
```
by
```py
for journal in sale_purchase_journals:
            if journal.type == "purchase" and query_results_payments_to_pay:
                query_results_to_pay[journal.id] = query_results_payments_to_pay[journal.id]
            else:
                query_results_to_pay[journal.id] = query_results_bills_to_pay[journal.id]
```
if it makes more sense to you.

For the testing part, there is already a test checking the values on the dashboard in odoo/addons/account/tests/test_account_journal_dashboard.py. I didn't modify it because the behavior did not change, We still check the total amount of the bills that are waiting for payment/overdue and those amount don't change based on whether you look at the moves or the amls. The only part of the behavior that is modified is the following : Previously, we were counting the number of moves that were involved in the total and now we count the number of amls that are involved. I can add a test for that part but it seems to me that if the total amount is right, then it means that every line that needed to be taken into account has been so testing the total is sufficient. Now that you know, I let it be your call.

File: addons/account/tests/test_account_payment_items.py [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+        # discount is the date_maturity
+        self.assertEqual(str(self.late_bill.line_ids[0].payment_date), '2023-03-01')
+        # Testing that the payment_date on the lines of a waiting for a payment
+        # move with a discount date that is not passed yet is the discount_date
+        self.assertEqual(str(self.discount_bill.line_ids[0].payment_date), '2023-03-20')
+        # Testing that the payment_date on the lines of a waiting for a payment
+        # move with a discount date that is already passed is the date_maturity
+        self.assertEqual(str(self.late_discount_bill.line_ids[0].payment_date), '2023-04-20')
+
+    def test_search_payment_date(self):
```
Comment (william-andre) [inline addons/account/tests/test_account_payment_items.py:124]: This is already a nice test :+1: , but what I believe would make it more readable is something like
```python
# bill:               due 2023-03-20
# late_bill:          due 2023-03-01
# discount_bill:      due 2023-03-25 discount 2023-03-20
# late_discount_bill: due 2023-04-01 discount 2023-03-10 
for today, search, expected in [
    ('2023-03-05', '2023-03-01', late_bill),
    ('2023-03-05', '2023-03-30', bill + late_bill + discount_bill + late_discount_bill),
    ...
]:
    with freeze_time(today):
        self.assertEqual(self.env['account.move.line'].search([
            ('payment_date', '=', search),
            ('partner_id', '=', self.private_partner_id.id),
        ]).move_id, expected)
```

Instead of the comments, a helper function that takes those as arguments would be even better
```python
bill = create_bill(due='2023-03-20')
late_bill = create_bill(due='2023-03-01')
discount_bill = create_bill(due='2023-03-25', discount='2023-03-20')
late_discount_bill = create_bill(due='2023-04-01', discount='2023-03-10')
```


This way all the information is at the same place and I think the comments become useless.

_____

In general, always [Keep it DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)

**Comment by william-andre** [comment]
Also, AL asked me to change the commit title from `[IMP] *:` to `[IMP] account:`. It is true that this is mainly a change for `account`. `*` would be used for a more frameworky change

**Comment by Megaaaaaa** [comment]
No change on the enterprise commit's name ? Or same but with account_report ?

**Comment by william-andre** [comment]
> No change on the enterprise commit's name ? Or same but with account_report ?

Same of course

File: addons/account/models/account_journal_dashboard.py [hunk 15]
```diff
@@ -664,6 +714,9 @@ def action_create_vendor_bill(self):
             'views': [[view_id, 'form']],
         }
 
+    def action_open_payment_items(self):
+        return self.env['ir.actions.actions']._for_xml_id('account.action_open_payment_items')
```
Comment (william-andre) [inline addons/account/models/account_journal_dashboard.py:718]: You still need to filter on the right journal, otherwise you see the lines of all journals.

I think this can be done easily by specifying `action_name` in the context, and using `open_action` instead (see `_select_action_to_open`)

File: addons/account/views/account_journal_dashboard_view.xml (hunk 5 above)
Comment (william-andre) [reply addons/account/views/account_journal_dashboard_view.xml:324]: So you can put it back :)

**Review by william-andre — COMMENTED** [review]
We're getting there! :smile:

**Comment by william-andre** [comment]
You also need to rebase (conflict in `hr_expense`) and fix the CI/style

**Review by william-andre — APPROVED** [review]
@robodoo r+

Thanks!
Sorry if my reviews were a bit hard sometimes, I had a lot to do and not a lot of time to waste :wink:

**Comment by william-andre** [comment]
@robodoo r+
